projects/observability/src/shared/components/topology/d3/layouts/graph-layout.ts

```diff
@@ -0,0 +1,76 @@
+import { RenderableTopology, RenderableTopologyNode, TopologyCoordinates, TopologyEdge, TopologyNode } from "../../topology";
+
+export class GraphLayout {
+
+  private levelToNodesMap = new Map<number, RenderableTopologyNode[]>();
+  private nodeToLevelMap = new Map<RenderableTopologyNode, number>();
+  private nodeCoOrdinatesMap: Map<TopologyNode, TopologyCoordinates> = new Map<TopologyNode, TopologyCoordinates>();
+
+  public layout(topology: RenderableTopology<TopologyNode, TopologyEdge>): void {
+
+    this.findRootNodes(topology);
+    this.fillNodeAndLevelMaps();
+    this.fillNodeCoOrdinatesMap();
+    this.setNodeCoOrdinates(topology);
+  }
+
+  private findRootNodes(topology: RenderableTopology<TopologyNode, TopologyEdge>) {
+    topology.nodes.forEach(node => {
+      if (node.incoming.length === 0) {
+        this.nodeToLevelMap.set(node, 0);
+        this.levelToNodesMap.has(0) ? this.levelToNodesMap.get(0)?.push(node) : this.levelToNodesMap.set(0, [node])
+      }
+    });
+  }
+
+  private fillNodeAndLevelMaps() {
+
+    const nodes = [...(this.levelToNodesMap.get(0) ?? [])];
+    const visited = new Set<TopologyNode>([...nodes.map(node => node.userNode)]);
+
+    while (nodes.length > 0) {
+      const parent = nodes[0];
+      const level = this.nodeToLevelMap.get(parent)! + 1;
+
+      parent.outgoing.forEach(edge => {
+        const child = edge.target;
+        if (!visited.has(child.userNode)) {
+          visited.add(child.userNode);
+          nodes.push(child);
+          this.nodeToLevelMap.set(child, level);
+          this.levelToNodesMap.has(level) ? this.levelToNodesMap.get(level)?.push(child) : this.levelToNodesMap.set(level, [child]);
+        }
+      });
+
+      nodes.shift();
+    }
+  }
+
+  private fillNodeCoOrdinatesMap() {
+
+    console.log(this.levelToNodesMap);
+
+    let curX = 1;
+
+    Array.from(this.levelToNodesMap.keys()).forEach(level => {
+      const nodes = this.levelToNodesMap.get(level)!;
+
+      let curY = 1;
```

Comment:
So we always start rendering at (1,1)? That doesn't seem desirable - but we can revisit. I think ideally we'd support centering (and if not centering, at least padding). If there were two nodes, right now this would render in the top left rather than right in the middle if I'm reading this right?

Reply:
Yeah you are right, it will, as of now, start render rendering from the top left, but we can bring each level into the center, as we know the height of that level and we know the height of the container, so we can just translate each of the nodes to the desired position

Reply:
Vertically center aligned the nodes

Reply:
we can make it configurable in this refactor [PR](https://github.com/hypertrace/hypertrace-ui/pull/1880)